icu4c/source/tools/genrb/reslist.cpp

```diff
@@ -1362,11 +1391,19 @@ SRBRoot::compactKeys(UErrorCode &errorCode) {
     }
     keys = fKeys + fKeysBottom;
     for (; i < keysCount; ++i) {
-        map[i].oldpos = (int32_t)(keys - fKeys);
+        int32_t keyOffset = static_cast<int32_t>(keys - fKeys);
+        while (!fIsPoolBundle && keysInUse.count(keyOffset) == 0) {
+            // Mark the unused key as deleted
+            while (*keys != 0) { *keys++ = 1; }
+            *keys++ = 1;
+            keyOffset = static_cast<int32_t>(keys - fKeys);
+        }
+        map[i].oldpos = keyOffset;
         map[i].newpos = 0;
         while (*keys != 0) { ++keys; }  /* skip the key */
         ++keys;  /* skip the NUL */
     }
+    U_ASSERT(keys == fKeys + fKeysTop);
```

Comment:
This will not be true if the last still-used key is followed by one or more filtered-out keys: The loop above stops after mapping the last used key.

Reply:
Great catch.  I'm able to reproduce the error.

The assertion needs to hold so that unused keys are overwritten.  The loop needs to be fixed.